### PR TITLE
Move device plugin related code to pkg/kubelet/cm/deviceplugin/.

### PR DESCRIPTION
**Pull request by jiayingz**
**What this PR does / why we need it**:
Move device plugin related code to pkg/kubelet/cm/deviceplugin/.
Inspired by PR 53258 proposed by @dolftax. The code structure looks much cleaner this way and it also makes future code refactoring easier.
While moving device_plugin_handler_test.go from pkg/kubelet/cm/ to pkg/kubelet/cm/deviceplugin/, we can no longer uses cm in its tests because that would cause a cycle dependency. To solve this problem, I moved the main cm GetResources functionality as well as part of the current device plugin handler Allocate functionality into a new device plugin handler function, GetDeviceRunContainerOptions(). This refactoring is also needed by another PR 51895 that moves device allocation into admission phase. Now device plugin handler Allocate() first checks whether there is cached device runtime state and only issues Allocate grpc call if there is no cached state available. The new GetDeviceRunContainerOptions() function simply returns device runtime config from the cached state. To support this change, extended the podDevices struct and checkpoint data structure with device runtime state.

We expect to clean up device plugin interface in follow up PRs. Currently both device plugin handler interface and manager interface are public. This is no longer necessary after this PR given that their code now live in the same package. We should clean up unnecessary public interfaces and functions, and/or consider to merge handler and manager code into a single layer.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
Part of https://github.com/kubernetes/kubernetes/issues/53170 but not fixes it.

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by RenaudWasTaken** [comment]
The only thing that "bugs" me is that @vishh wanted to have a separate package that does not depend on kubernetes.

IMHO as long as we keep the manager interface this change is good!

**Comment by lichuqiang** [comment]
Wondering if you could separate the change in **pod_devices** and merge that part first, I think it may have impact on many upcoming PRs.

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (lichuqiang) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: It would be a bit tricky to see whether we get real containerDevices or not by returning a DeviceRunContainerOptions item. I would prefer to return a reference here.
In addition,  transmission for structs may have impact on performance

File: pkg/kubelet/cm/container_manager_linux.go
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/container_manager_linux.go]: Nit: if is unneeded appending nil lists to nil lists returns nil

**Comment by RenaudWasTaken** [comment]
Looking into this a bit more I'm very confused by this PR, it seems to be doing a lot more than just moving code from one place to another.

Moving the files has the unintended consequence that it prevents looking at the code diff correctly and makes this whole PR a lot more complicated to read.

Can you split this PR into multiple commits ? This would make it a lot more easier to review.

**Comment by RenaudWasTaken** [comment]
> While moving device_plugin_handler_test.go from pkg/kubelet/cm/ to pkg/kubelet/cm/deviceplugin/, we can no longer uses cm in its tests because that would cause a cycle dependency. 

The mentioned test seems to not be a unit test which would explain why you had so many problems trying to move it.
That test is testing a method of `containerManagerImpl` in `device_plugin_handler` which should have been unit tested in `container_manager_linux_test.go` with a stub `DevicePluginHandler`.
This would have allowed to send and test different combinations of `AllocationResponses` to `GetResources`.

> To solve this problem, I moved the main cm GetResources functionality as well as part of the current device plugin handler Allocate functionality into a new device plugin handler function, GetDeviceRunContainerOptions(). 

Can we break this change into another PR?
As mentioned before the whole PR is not very reviewer friendly and the introduced changes have a pretty important scope (checkpoint + new struct + new tests).

The mentioned test is currently disabled so we can just comment it out and re-introduce it in the new PR.

> This refactoring is also needed by another PR #51895 that moves device allocation into admission phase.

Moving this code into another PR shouldn't impact PR #51895 anymore.
Anyways you also mentioned that we needed coordination, I believe we should wait until we have a definite architecture design document before merging anything.

> Now device plugin handler Allocate() first checks whether there is cached device runtime state and only issues Allocate grpc call if there is no cached state available. The new GetDeviceRunContainerOptions() function simply returns device runtime config from the cached state. To support this change, extended the podDevices struct and checkpoint data structure with device runtime state.

See my comment about this completely widening the scope of this PR

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (vishh) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: A pod's resource cannot be updated once admitted by the API server. I'd much rather fail loudly in this case (not crashloop) than handle it.

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (vishh) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: Will devices get stranded if Allocate fails? IIUC, devices are removed from Allocatable pool prior to actual allocation.

Comment (vishh) [reply pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: nvm. I see the note on `updateAllocatedDevices` below

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (vishh) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: This can result in a race - Imagine a pod getting rejected by the kubelet because it temporarily stranded resources that it has previously allocated to a pod it had rejected because Allocation failed.

Comment (vishh) [reply pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: May be run an update before every Allocation and ensure that the update is cheap?

**Review by vishh — APPROVED** [review]
/lgtm
/approve

**Review by jiayingz — COMMENTED** [review]
Thanks a lot for the reviews! PTAL.

**Review by RenaudWasTaken — CHANGES_REQUESTED** [review]
Would like to see this PR at least broken down into multiple commits.
This would allow us to review much more confidently this PR based on a diff rather than try to find bugs in a +701 −447 commit.

If not broken into multiple PRs whose scope are:
- Only to move the code to  pkg/kubelet/cm/deviceplugin/ as it is defined in the title
- Change the behavior of cm

**Comment by jiayingz** [comment]
@RenaudWasTaken Thanks a lot for your suggestions! I did consider to break this PR into multiple ones and consulted with @vishh yesterday. He suggested to keep the PR as is, given that we are already in the middle of review and this PR is needed for some ongoing PRs. I have described the scope of the change in both PR and commit descriptions. Hopefully that will help future retrospection. Most of the changes included in this PR are code move around, although unfortunately git is not smart enough to recognize that. I generally don't like to have a commit that leaves tests or build in broken state, but I see your point that this could help speed up code review. I will keep that in mind next time. As of avoiding introducing bugs through this PR change, I have made sure both unit tests and e2e tests continued to pass with the PR.

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (vishh) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: One more consideration for the future: Why not allocate multiple resources in parallel? Even if it's serial, kubelet needs to deal with failed allocations (currently handled via `updateAllocations` method).

**Review by vishh — APPROVED** [review]
/lgtm
/approve

**Comment by RenaudWasTaken** [comment]
> although unfortunately git is not smart enough to recognize that. 

Which is why I suggested to break this PR in multiple commits.
Every commit does not need to compile just the last one.

> Most of the changes included in this PR are code move around,  I generally don't like to have a commit that leaves tests or build in broken state, but I see your point that this could help speed up code review. I will keep that in mind next time. As of avoiding introducing bugs through this PR change, I have made sure both unit tests and e2e tests continued to pass with the PR.

It's not even about speeding code review. We can't reasonably expect people to review 1000 lines of changes in a single commit.

Can we at least discuss this in tomorrow's meeting before merging this PR ?

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: nit: keep new line :)

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: Looks like this won't continue if needed = 0

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: Question for the future: How do we expect to report used / unused devices ?

With the current model if devices get stranded they would be advertised as taken right ?
In that case wouldn't a model that cleans up stranded devices be a better one ?

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: I might be understanding this wrong but this part of the code seems to suggest that you might not get the same devices between reboots right ?

Wouldn't we want a container to keep the exact same devices between reboots ?

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: Same comment about Big lock here

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: Same comment about Big lock here

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: I'm a bit surprised by the usage of the Mutex here. Usually you would want to keep critical section to a minimum.
We are more in a big lock situation here.

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (RenaudWasTaken) [inline pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: This needs to be out of the critical section. Performance is going to be very impacted

**Review by RenaudWasTaken — CHANGES_REQUESTED** [review]
The locking strategy doesn't seem like a good idea.
Critical sections are huge.

We should at least address file write in a critical section in this PR.

We can address the complex locking mechanism in future PRs (discussion could happen in the design document).

The complex locking mechanism is not going to be an easy problem to solve because it relates to the global architecture of the device plugin

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (jiayingz) [reply pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: Right now we don't expect it is a common case that a container require resources from multiple device plugins. If that happens in the future, I agree we should consider this. Added a comment about this.

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (jiayingz) [reply pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: I think this is needed to synchronize read/write access on h.podDevices. Also this part shouldn't be expensive.

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (jiayingz) [reply pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: Good point. I added this part when I thought we need to support resource downsize and upsize, but then learned from Vish that a pod's resource requirements can't change as of today. Simplified this part.

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (jiayingz) [reply pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: that part now moves to devicesToAllocate().

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (jiayingz) [reply pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: done.

File: pkg/kubelet/cm/deviceplugin/device_plugin_handler.go
Comment (jiayingz) [reply pkg/kubelet/cm/deviceplugin/device_plugin_handler.go]: Good point, I added the code to restore h.allocatedDevices upon allocation failures.

**Comment by vishh** [comment]
I feel this PR is a net positive from where the code is today. Do you agree @RenaudWasTaken ? If yes, then let's merge this PR and continue improving the code?

**Comment by RenaudWasTaken** [comment]
Agreed :)

**Pull request merged**
